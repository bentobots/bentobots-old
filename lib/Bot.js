'use strict';

/**
 * A bot (node) in the graph.
 * @constructor
 * @param {string} id - The unique identifier of the bot
 * @param {reference} component - The script that the bot uses
 * @param {inputs} object - A key/value object representing the values passed to the bot
 * @param {metadata} object
 */
function Bot(id, component) {
  var inputs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  var metadata = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};

  this.id = id;
  this.component = component;
  this.inputs = inputs;
  this.metadata = metadata;
  this.outputs = component.outputs;
}

Bot.prototype.work = function () {
  var _this = this;

  var inputs = {};
  Object.keys(this.inputs).forEach(function (k) {
    var value = _this.inputs[k];
    if (typeof value === 'string' || value instanceof String) {
      var patternMatch = value.match(/([A-Z][A-Za-z0-9]+)>([A-Z]+)/);
      if (patternMatch) {
        // if it matches Process1>PORT, then it's an alias
        inputs[k] = _this.graph.data[value.split('>')[0]][value.split('>')[1]];
      } else {
        inputs[k] = value;
      }
    } else {
      inputs[k] = value;
    }
  });
  this.graph.data[this.id] = this.outputs = this.component.implementation(inputs);
};

module.exports = Bot;