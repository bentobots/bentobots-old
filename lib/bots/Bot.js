'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Bot = function () {
  function Bot(id, component) {
    var inputs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    var metadata = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};

    _classCallCheck(this, Bot);

    this.id = id;
    this.component = component;
    this.inputs = inputs;
    this.metadata = metadata;
    this.outputs = { SUM: 0 };
  }

  _createClass(Bot, [{
    key: '_outputs',
    value: function _outputs() {
      return this.outputs;
    }
  }, {
    key: 'work',
    value: function work() {
      var _this = this;

      // console.log(`running ${this.id}`)
      var inputs = {};
      Object.keys(this.inputs).forEach(function (k) {
        var value = _this.inputs[k];
        if (typeof value === 'string' || value instanceof String) {
          inputs[k] = _this.graph.data[value.split('>')[0]][value.split('>')[1]];
        } else {
          inputs[k] = value;
        }
      });
      // console.log(`working: ${this.id}`)
      this.graph.data[this.id] = this.outputs = this.component(inputs);
    }

    // addInPort({ id, validTypes }) {
    //   {
    //     id: id,
    //     validTypes: validTypes
    //   }
    // }

    // static registerClass (key, constructor) {
    //   constructor.type = key
    //   Bot.classes[key] = constructor
    // }

  }]);

  return Bot;
}();

module.exports = Bot;