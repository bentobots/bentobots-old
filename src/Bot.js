class Bot {
  constructor (id, component, inputs = {}, metadata = {}, visible=true) {
    this.id = id
    this.component = component
    this.outputs = component.outputs
    this.inputs = inputs
    this.metadata = metadata
    this.visible = visible
  }

  work () {
    // console.log(`running ${this.id}`)
    const inputs = {}
    Object.keys(this.inputs).forEach(k => {
      const value = this.inputs[k]
      if (typeof value === 'string' || value instanceof String) {
        const patternMatch = value.match(/([A-Z][A-Za-z0-9]+)>([A-Z]+)/)
        if (patternMatch) { // if it matches Process1>PORT, then it's an alias
          inputs[k] = this.graph.data[value.split('>')[0]][value.split('>')[1]]
        } else {
          inputs[k] = value
        }
      } else {
        inputs[k] = value
      }
    })
    // console.log(`working: ${this.id}`)
    this.graph.data[this.id] = this.outputs = this.component.implementation(inputs)
  }

  // addInPort({ id, validTypes }) {
  //   {
  //     id: id,
  //     validTypes: validTypes
  //   }
  // }

  // static registerClass (key, constructor) {
  //   constructor.type = key
  //   Bot.classes[key] = constructor
  // }
}

module.exports = Bot
