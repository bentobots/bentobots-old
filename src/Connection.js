function Connection({ srcBot, srcPort, tgtBot, tgtPort, graph } = {}) {
  this.srcBot = srcBot
  this.srcPort = srcPort
  this.tgtBot = tgtBot
  this.tgtPort = tgtPort
  this.graph = graph
}

Connection.prototype.isValid = function() {
  if (!this.graph) return false
  if (srcBot === tgtBot) return false
  if (!tgtBot.inPorts.includes(tgtPort)) return false
  if (!srcBot.outPorts.keys.includes(outPort)) return false
  if (!srcBot.outPorts.keys.includes(outPort)) return false
  const validTypes = tgtBot.inPorts.keys[tgtPort].validTypes
  if (validTypes && !validTypes.includes(srcBot.outPorts.keys[srcPort].type)) return false
  return true
}

export default Connection
