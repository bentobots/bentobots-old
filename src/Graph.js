import Immutable from 'immutable'
import memoize from 'memoizee'
import ASQ from 'asynquence'
import topolysis from 'topolysis'

class Graph {

  constructor ({name = 'My Graph'} = {}) {
    this.name = name
    this.bots = Immutable.List()
    this.connections = Immutable.List()
    // this._graph = new graphlib.Graph({directed: true})
    this.graph = {}
    this.sorted = []
  }

  addBot (bot) {
    if (!bot) throw new Error('Bot not given')
    if (bot.graph) throw new Error('Bot was already added to graph')
    if (this.bots.find(b => b.id === bot.id)) {
      throw new Error('Bot with that id already in graph')
    }
    bot.graph = this
    this.graph[bot.id] = []
    this.bots = this.bots.push(bot)
    this.sort()
  }

  removeBot (bot) {
    delete this.graph[bot.id]
    this.bots = this.bots.filterNot(b => b === bot)
    this.sort()
  }

  getBotById (id) {
    return this.bots.find(b => b.id === id)
  }

  addConnection (conn) {
    if (conn.graph) throw new Error('Connection was already added to a graph')

    this.graph[conn.srcBot.id].push(conn.tgtBot.id)

    conn.graph = this
    this.connections = this.connections.push(conn)
    this.sort()
  }

  removeConnection (conn) {
    const i = this.graph[conn.srcBot.id].indexOf(conn.tgtBot.id)
    if (i != -1) { this.graph[conn.srcBot.id].splice(i, 1) }

    this.connections = this.connections.filterNot(c => c === conn)
    this.sort()
  }

  sort () {
    let toReverse = []
    for (const ids of topolysis(this.graph)) {
      toReverse.push(ids)
    }
    this.sorted = toReverse.reverse()
    // toReverse.reverse().forEach(ids => {
    //   // console.log(`RUNNING BATCH SEQUENTIALLY...... ${ids}`)
    //   // ids.forEach(id => bots.find(b => b.id === id).work())
    //   console.log(`RUNNING BATCH IN PARALLEL...... ${ids}`)
    //   ASQ().all(...ids.map(wrap))// .val(function () { console.log(arguments) })
    // })
  }

}

export default Graph
