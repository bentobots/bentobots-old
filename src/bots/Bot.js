class Bot {

  constructor (id, component, inputs = {}, metadata = {}) {
    this.id = id
    this.component = component
    this.inputs = inputs
    this.metadata = {}
    // this.outputs = outputs
  }

  work () {
    // console.log(`running ${this.id}`)
    const inputs = {}
    Object.keys(this.inputs).forEach(k => {
      const value = this.inputs[k]
      if (typeof value === 'string' || value instanceof String) {
        inputs[k] = this.graph.data[value.split('>')[0]][value.split('>')[1]]
      } else {
        inputs[k] = value
      }
    })
    // console.log(`working: ${this.id}`)
    this.graph.data[this.id] = this.component(inputs)
  }

  // addInPort({ id, validTypes }) {
  //   {
  //     id: id,
  //     validTypes: validTypes
  //   }
  // }

  // static registerClass (key, constructor) {
  //   constructor.type = key
  //   Bot.classes[key] = constructor
  // }

}

module.exports = Bot
