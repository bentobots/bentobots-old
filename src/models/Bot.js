/**
 * A bot (node) in the graph.
 * @constructor
 * @param {string} id - The unique identifier of the bot
 * @param {reference} component - The script that the bot uses
 * @param {inputs} object - A key/value object representing the values passed to the bot
 * @param {metadata} object
 */
function Bot(id = x`id`, component = x`component`, inputs = {}, metadata = {}) {
  this.id = id
  this.component = component
  this.inputs = inputs
  this.metadata = metadata
  this.outputs = component.outputs
}

Bot.prototype.work = function() {
  const inputs = {}
  Object.keys(this.inputs).forEach(k => {
    const value = this.inputs[k]
    if (typeof value === 'string' || value instanceof String) {
      const patternMatch = value.match(/([A-Z][A-Za-z0-9]+)>([A-Z]+)/)
      if (patternMatch) { // if it matches Process1>PORT, then it's an alias
        inputs[k] = this.graph.data[value.split('>')[0]][value.split('>')[1]]
      } else {
        inputs[k] = value
      }
    } else {
      inputs[k] = value
    }
  })
  this.graph.data[this.id] = this.outputs = this.component.implementation(inputs)
}

const x = p => { throw new Error(`Missing parameter: ${p}`) }

module.exports = Bot
