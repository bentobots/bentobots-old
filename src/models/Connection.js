/**
 * A connection (edge) between two bots (edges) in the graph
 * @constructor
 * @param {Bot} srcBot - The source bot to take data from
 * @param {string} srcPort -Which output from the source bot to pull from
 * @param {Bot} tgtBot - The target bot to send data to
 * @param {string} srcPort - Which input on the target bot to push to
 */
function Connection({ srcBot, srcPort, tgtBot, tgtPort } = {}) { // , graph
  this.srcBot = srcBot
  this.srcPort = srcPort
  this.tgtBot = tgtBot
  this.tgtPort = tgtPort
  // this.graph = graph
}

Connection.prototype.isValid = function() {
  if (!this.graph) return false
  if (srcBot === tgtBot) return false
  if (!tgtBot.inPorts.includes(tgtPort)) return false
  if (!srcBot.outPorts.keys.includes(outPort)) return false
  if (!srcBot.outPorts.keys.includes(outPort)) return false
  const validTypes = tgtBot.inPorts.keys[tgtPort].validTypes
  if (validTypes && !validTypes.includes(srcBot.outPorts.keys[srcPort].type)) return false
  return true
}

export default Connection
