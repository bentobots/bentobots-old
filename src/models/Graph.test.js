import Graph from './Graph'
import Bot from './Bot'
import Connection from './Connection'

const botMock = {
  inputs: []
}

test('can be created, with default name', () => {
  const graph = new Graph()
  expect(graph.name).toMatchSnapshot()
})

test('can multiple bots with .addBots', () => {
  const graph = new Graph()
  // graph.addBots([bot1, bot2])
  expect(graph.bots.size).toEqual(2)
})

// const graphWithBots = new Graph()
// const first = new Bot('first', () => {})
// const second = new Bot('second', () => {})

// graphWithBots.addBots([first,second])

// it('can get a bot by id', () => {
//   expect(graphWithBots.getBotById('second')).toEqual(second)
// })

// it('cannot add the same bot twice', () => {
//   expect(() => {
//     graphWithBots.addBot(second)
//   }).toThrow(new Error('Bot was already added to graph'))
// })

// it('cannot add a bot with the same name', () => {
//   const third = new Bot('second', () => {})
//   expect(() => {
//     graphWithBots.addBot(third)
//   }).toThrow(new Error("Bot with id 'second' already in graph"))
// })

// it('can remove a bot', () => {
//   const graphWithBotsClone = graphWithBots
//   expect(graphWithBotsClone.bots.toArray().length).toEqual(2)
//   graphWithBotsClone.removeBot(second)
//   expect(graphWithBotsClone.bots.toArray().length).toEqual(1)
// })

// it('can add a connection', () => {
//   const graph = new Graph()
//   const bot1 = new Bot('first', () => {})
//   const bot2 = new Bot('second', () => {})
//   graph.addBots([bot1, bot2])
//   const connection = new Connection({ srcBot: bot1, srcPort: "OUT", tgtBot: bot2, tgtPort: "IN" })
//   graph.addConnection(connection)
// })

// it('can remove a connection')
