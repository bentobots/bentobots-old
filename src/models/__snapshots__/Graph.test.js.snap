// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be created, with default name 1`] = `"Untitled Graph"`;
